src/ahp/indexer.rs

```diff
@@ -112,21 +115,27 @@ impl<F: PrimeField> AHPForR1CS<F> {
         let index_time = start_timer!(|| "AHP::Index");
 
         let constraint_time = start_timer!(|| "Generating constraints");
-        let mut ics = IndexerConstraintSystem::new();
-        c.generate_constraints(&mut ics)?;
+        let ics = ConstraintSystem::new_ref();
+        c.generate_constraints(ics.clone())?;
```

Comment:
I think you should invoke `ics.set_mode(SynthesisMode::Setup)` here. (Maybe I should rename that to `SynthesisMode::Index`?)